docs/_posts/2024-02-27-whats-new-in-cdi41.md

```diff
@@ -0,0 +1,115 @@
+
+title: "What's new in CDI 4.1?"
+author: ladt
+tag: development-tips
+---
+
+If you're following the activity on the [CDI repository](https://github.com/jakartaee/cdi/), or in Jakarta EE in general, you might have noticed that the CDI 4.1 release is coming soon.
+In this short post, we'll take a look at what's new in the wonderful world of CDI.
+
+## Method Invokers
+
+The most important (or at least the biggest) addition to CDI is _method invokers_.
+It is not terribly interesting for application developers, but it should be very useful for people who build CDI-based frameworks.
+
+A method invoker is, unsurprisingly, an object that allows invoking a method indirectly.
+It implements the `jakarta.enterprise.invoke.Invoker` interface, which is very simple:
+
+```java
+interface Invoker<T, R> {
+    R invoke(T instance, Object[] arguments) throws Exception;
+}
+```
+
+CDI extensions can create an invoker for methods (including `static` methods) declared on managed beans.
+For example, let's say the application contains this bean:
+
+```java
+@Dependent
+public class MyService {
+    public String hello(String name) {
+        return "Hello " + name + "!";
+    }
+}
+```
+
+A hypothetical CDI-based framework can decide that it might need to invoke the method at runtime.
+During deployment, the CDI extension provided by the framework would use
+
+- `ProcessManagedBean.createInvoker()`, if using a Portable Extension
+- `InvokerFactory.createInvoker()`, if using a Build Compatible Extension
+
+to obtain an `InvokerBuilder`. Calling `build()` on it will produce
+
+- an `Invoker` directly, if using a Portable Extension
+- an `InvokerInfo` that materializes as an `Invoker` at application runtime, if using a Build Compatible Extension
+
+The `Invoker` then needs to be stored somewhere until the framework actually needs it.
+When the time comes to invoke the method, the framework takes the `Invoker`, obtains an instance of `MyService` and all the parameter values, and calls
+
+```java
+invoker.invoke(myService, new Object[] { "world" })
+```
+
+The return value is `"Hello world!"`, as expected.
+
+### Lookups
+
+It is fairly common that some of the parameters the method accepts are CDI beans.
+It is possible to configure the `InvokerBuilder` to resolve and instantiate a correct value for such parameter automatically.
+In fact, this is possible also for the instance on which the method should be invoked.
+The `InvokerBuilder` has methods `withInstanceLookup()` and `withArgumentLookup()` to configure that.
+
+Returning back to the example above, if the CDI extension called `withInstanceLookup()` before calling `InvokerBuilder.build()`, it wouldn't have to obtain the instance of `MyService` manually.
+The invoker would do it automatically.
+The invocation would just use `null` as the instance value:
+
+```java
+invoker.invoke(null, new Object[] { "world" })
+```
+
+The result is still the same: `"Hello world!"`.
+
+## Other Improvements
```

Comment:
Shouldn't we also mention the split of EE parts of CDI specification (and respective TCK bits) into the umbrella spec?
Probably not under `Other Improvements`, maybe in a separate `Housekeeping` section or something like that?

Reply:
Also good point.

Reply:
Done.